### PR TITLE
feat(mjolnir): LOG_WARN counters rather than single occurrences, add statsd support to tile build

### PR DESCRIPTION
**Pull request by nilsnolde**
fixes https://github.com/valhalla/valhalla/issues/5977

claude did the base changes, then I spent a rather long time changing the changes:) the basic gist is:

- `LOG_WARN` in mjolnir's hot loops are changed to `LOG_DEBUG` and a `LOG_WARN` for a metric summary after the loop(s)
- included `statsd` to mjolnir for the `LOG_WARN` summary counters: I plan to hook that up to my own Grafana instance and give other maintainers read access; could've been done separately actually.. anyways, I think it's fine now, I'll update more statsd stuff maybe in the future

a tile build for berlin shows this

```
2026-03-24 21:05:15.729075617 [INFO] EdgeInfo Count = 955417
2026-03-24 21:05:15.729079194 [TIMING] src/mjolnir/graphbuilder.cc::Build took 1s
2026-03-24 21:05:15.729083271 [WARN] [build] 2 ways with invalid level tags
2026-03-24 21:05:15.729903878 [INFO] Enhancing local graph...
```

### CODE REVIEW

File: src/mjolnir/complexrestrictionbuilder.cc [hunk 1]
```diff
@@ -12,7 +12,7 @@ namespace mjolnir {
 // Set the via edge ids used by this complex restriction.
 void ComplexRestrictionBuilder::set_via_list(const std::vector<GraphId>& via_list) {
   if (via_list.size() > kMaxViasPerRestriction) {
-    LOG_WARN("Tried to exceed max vias per restriction: " + std::to_string(via_list.size()));
+    LOG_DEBUG("Tried to exceed max vias per restriction: " + std::to_string(via_list.size()));
```
Comment (nilsnolde) [inline src/mjolnir/complexrestrictionbuilder.cc:15]: these 2 logs are kinda dead code IIUC. we already count and log in restrictionbuilder.cc

**Comment by nilsnolde** [comment]
actually I wonder if we couldn't do a singleton pattern here for the build stats. instead of passing them through to every caller.. I'll check

**Comment by nilsnolde** [comment]
claude agrees;) happy to see that, shows again the limitations of AI. or maybe my laziness to force it to review itself after a long session like I had, manually changing things around. too lazy now to find out though as well hahah

File: src/mjolnir/edgeinfobuilder.cc [hunk 2]
```diff
@@ -125,14 +129,15 @@ std::ostream& operator<<(std::ostream& os, const EdgeInfoBuilder& eib) {
   auto ei = eib.ei_;
   uint32_t name_count = eib.name_info_list_.size();
   if (name_count > kMaxNamesPerEdge) {
-    LOG_WARN("Exceeding max names per edge: " + std::to_string(name_count));
+    LOG_DEBUG("Exceeding max names per edge: " + std::to_string(name_count));
```
Comment (nilsnolde) [inline src/mjolnir/edgeinfobuilder.cc:130]: this shouldn't fire actually (guarded before getting here), it's also dead code I think

File: src/mjolnir/graphbuilder.cc [hunk 3]
```diff
@@ -660,23 +662,14 @@ void BuildTileSet(const std::string& ways_file,
           if (!use_admin_db)
             dor = w.drive_on_right();
 
-          // Validate speed. Set speed limit and truck speed.
+          // Speed values are already clamped to kMaxAssumedSpeed (140) in OSMWay setters.
```
Comment (nilsnolde) [inline src/mjolnir/graphbuilder.cc:664]: removes some code

File: src/mjolnir/graphtilebuilder.cc [hunk 4]
```diff
@@ -722,7 +724,8 @@ uint32_t GraphTileBuilder::AddEdgeInfo(const uint32_t edgeindex,
     for (const auto& name : names) {
       // Stop adding names if max count has been reached
       if (name_count == kMaxNamesPerEdge) {
-        LOG_WARN("Too many names for edgeindex: " + std::to_string(edgeindex));
+        LOG_DEBUG("Too many names for edgeindex: " + std::to_string(edgeindex));
+        build_stats::get().increment(build_stats::kExceededMaxNames);
```
Comment (nilsnolde) [inline src/mjolnir/graphtilebuilder.cc:728]: I had claude deeply investigate the 5 occurrences here and seems all fine. I guess one or some of these might prevent the others to even ever get triggered by early limiting the max names

File: src/mjolnir/shortcutbuilder.cc [hunk 5]
```diff
@@ -588,26 +589,27 @@ std::pair<uint32_t, uint32_t> AddShortcutEdges(GraphReader& reader,
...
-  // Log a warning (with the node lat,lon) if the max number of shortcuts from a node
-  // is exceeded. This is not serious (see NOTE above) but good to know where it occurs.
+  // Log if the max number of shortcuts from a node is exceeded.
+  // This is not serious (see NOTE above) but good to know where it occurs.
   if (shortcut_count > kMaxShortcutsFromNode) {
-    PointLL ll = tile->get_node_ll(start_node);
-    LOG_WARN("Exceeding max shortcut edges from a node at LL = " + std::to_string(ll.lat()) + "," +
-             std::to_string(ll.lng()));
+    [[maybe_unused]] PointLL ll = tile->get_node_ll(start_node);
+    LOG_DEBUG("Exceeding max shortcut edges from a node at LL = " + std::to_string(ll.lat()) + "," +
```
Comment (nilsnolde) [inline src/mjolnir/shortcutbuilder.cc:596]: further down we increment max_exceeded counter and print a LOG_WARN summary plus add the relevant statsd counter

File: src/mjolnir/valhalla_add_elevation.cc [hunk 6]
```diff
@@ -116,6 +117,8 @@ int main(int argc, char** argv) {
     return EXIT_FAILURE;
   }
 
+  build_stats::get().reset();
   ElevationBuilder::Build(config, tile_ids);
+  build_stats::get().report(config, false);
```
Comment (nilsnolde) [inline src/mjolnir/valhalla_add_elevation.cc]: for this executable we only print logs but not emit statsd

File: valhalla/mjolnir/util.h [hunk 7]
```diff
@@ -90,6 +91,76 @@ inline std::string to_string(BuildStage stg) {
   return (i == BuildStageStrings.cend()) ? "null" : i->second;
 }
 
+// Counters for warnings that fire per-item in hot loops during tile building.
+// Instead of logging each occurrence (which produces millions of lines on planet builds),
+// we accumulate counts and log a summary at the end. Full per-item detail is still
```
Comment (nilsnolde) [inline valhalla/mjolnir/util.h:98]: this is all a bit more contrived, I wanted to be as safe as we can be, by linking build_stats counters with their "meta" implementation with a static_assert, to make future extensions safer

File: valhalla/mjolnir/util.h [hunk 8]
```diff
@@ -90,6 +91,76 @@ inline std::string to_string(BuildStage stg) {
...
+      {"build.access_tags_not_found", "edges with access tags not found"},
+      {"build.unrecognized_hov_type", "ways with unrecognized HOV type"},
+      {"build.tag_parse_error", "tag parse errors"},
+      {"build.exceeded_max_vias", "restrictions exceeding max vias"},
+      {"build.exceeded_max_shortcut_edges", "nodes exceeding max shortcut edges"},
+  };
+
+  static_assert(std::size(meta) == kCount, "build_stats::meta and counter enum are out of sync");
+
+  std::atomic<uint32_t> counters[kCount]{};
```
Comment (nilsnolde) [inline valhalla/mjolnir/util.h]: `atomic` bcs multiple threads can increment concurrently in some stages

**Comment by nilsnolde** [comment]
this has been lots of back & forth.. but I think it's now fairly nice. also made sure the counter increments make sense, as you'll see, there's are lots of counters which get called multiple times (I didn't go through all the paths manually, but had claude do a guided deep investigation on the suspicious ones).

**Comment by nilsnolde** [comment]
I'll first try this out and see how it lands in Grafana.

File: valhalla/mjolnir/util.h [hunk 9]
```diff
@@ -90,6 +91,76 @@ inline std::string to_string(BuildStage stg) {
...
+    ++counters[c];
+  }
+
+  static build_stats& get() {
+    static build_stats instance;
+    return instance;
+  }
+
+  void reset();
+  void report(const boost::property_tree::ptree& config, const bool emit_statsd = true) const;
```
Comment (kevinkreiser) [inline valhalla/mjolnir/util.h]: could whether or not to emit statsd simply be controlled by its existence in config (like the service is)? my thinking is tile building config is often different from service running config anyway so we can let that determine it, but i didnt think too hard about it admittedly

File: valhalla/mjolnir/util.h [hunk 10]
```diff
@@ -90,6 +91,76 @@ inline std::string to_string(BuildStage stg) {
...
+  void increment(counter c) {
+    ++counters[c];
+  }
+
+  static build_stats& get() {
+    static build_stats instance;
+    return instance;
+  }
+
+  void reset();
```
Comment (kevinkreiser) [inline valhalla/mjolnir/util.h]: having a resettable singleton seems strange to me. kind of the idea of a singleton is that you can use it but you cant mess with its internal state directly exactly because its shared. i know we can take care not to screw that up but maybe there is a better pattern to use? i guess the idea is that we dont want to pass it around everywhere, and a singleton handles that by being global to the namespace, but yeah having a global that anyone could reset randomly is a bit unfortunate. i dont have a concrete suggestion at the moment, just seems strange. maybe once i read a bit more ill change my mind. maybe another question, when do we need to reset it? the one that comes to mind for me is testing but perhaps we could solve that in other ways?

**Comment by kevinkreiser** [comment]
can you explain more why we want to accumulate things and report on them at the very end rather than a bit more timely as they occur? i totally get we dont want reams of log noise, so counting and logging makes sense to me! but i feel like its quite inconvenient to accumulate all of it for the whole build and log only at the end. for statsd i feel even more like the stats should be emitted as they happen considering statsd itself already has logic in place to batch this information and emit it on some interval. we're kind of duplicating that logic.

so yeah, i think my ideal would be, for statsd to emit the stats as they happen but then to log the final count at the end of each phase of the build so we dont have to get to the end to see those messages. its better to have them in the phase of the build where they match the phases current context you know what i mean?

anyway thats just my preference!

**Comment by nilsnolde** [comment]
just skimming your suggestions, I think I agree with most points. I didn't really think about the "when it's emitted/logged" in the processing. 

didn't try it yet either, lots of things to wire up before. but in a few hours I'll be there:)

**Comment by kevinkreiser** [comment]
no rush on my part and thank you again of course for the much needed clean up!

File: valhalla/mjolnir/util.h (hunk 9 above)
Comment (nilsnolde) [reply valhalla/mjolnir/util.h]: > could whether or not to emit statsd simply be controlled by its existence in config (like the service is)

it should be. that bool is only there bcs another executable (which we(/I) don't want to emit to statsd from), valhalla_add_elevation.cc I think.

**Comment by nilsnolde** [comment]
I'm more excited about statsd support for tile build than clean up tbh :smile:

**Comment by nilsnolde** [comment]
> can you explain more why we want to accumulate things and report on them at the very end rather than a bit more timely as they occur?

@kevinkreiser I agree, we should report after every stage. but I don't think we should emit statsd _during_ loops as a general pattern in the tile build. those happen during multiple threads and could be pretty noisy. while it'd be more in the spirit of statsd I'd rather emit them as well after each stage built. in monitoring I guess we'd be more interested in seeing the stats of the last tile build rather than following a live one.

**Comment by kevinkreiser** [comment]
threading is no issue for the statsd client but also if you feel strongly about accumulating its fine with me!

the only strawman i can think of for simply sending them live is if we can somehow gain something of seeing frequency of observations over time. eg. if we emit a metric for every tile written in the build stage and we see wow at the beginning of the build stage we are building tiles at like 10 per second but then there is an incredibly long tail at the end where we only emit 1 every 30 seconds, it would clue us into a problem we've had in the past where we dont spread out the work evenly enough to make maximal use of parallelism. its a very contrived example but i think thats the only thing that comes to mind that we would miss out on. stuff thats a problem but is only discoverable when also viewed on the time axis

anyway like i said i dont feel strongly about it, just trying to explain possible differences in the approach!

**Comment by nilsnolde** [comment]
The example makes sense. Not necessarily for the scope of this PR, where it's mostly "smelly" things and I can't think of any time series use case for those.

I'll definitely follow up with more build stats, basic graph metrics, timing etc.

File: valhalla/mjolnir/util.h (hunk 10 above)
Comment (nilsnolde) [reply valhalla/mjolnir/util.h]: yeah, the resetting was still kinda unreviewed AI slop. I guess it figured it might be called multiple times within a process. which is also true: the gurka tests do that, but there's none (and I don't think it's really necessary to have some).

File: scripts/valhalla_build_config [hunk 11]
```diff
@@ -439,7 +439,7 @@ config = {
         "host": Optional(str),
         "port": 8125,
         "prefix": "valhalla",
-        "batch_size": Optional(int),
+        "batch_size": 500,
```
Comment (nilsnolde) [inline scripts/valhalla_build_config:442]: we also default it in c++ code, but can't get rid of that now

File: src/mjolnir/util.cc [hunk 12]
```diff
@@ -692,6 +693,12 @@ bool build_tile_set(const boost::property_tree::ptree& original_config,
   std::string new_to_old_bin = tile_dir + new_to_old_file;
   std::string old_to_new_bin = tile_dir + old_to_new_file;
 
+  // Snapshot for per-stage delta reporting
+  uint32_t stats_snapshot[build_stats::kCount]{};
+  auto log_stage = [&stats_snapshot, &config](BuildStage stage) {
+    build_stats::get().log_stage(stage, stats_snapshot, config);
+  };
```
Comment (nilsnolde) [inline src/mjolnir/util.cc:701]: changed it to logging each stage after it's done.

File: src/mjolnir/util.cc [hunk 13]
```diff
@@ -899,5 +914,46 @@ TileManifest TileManifest::ReadFromFile(const std::string& filename) {
...
+    uint32_t delta = current - snapshot[i];
+    if (delta > 0) {
+      LOG_WARN(std::format("[{}] {} {}", stage_name, delta, meta[i].log_label));
+      // e.g. "build.exceeded_max_speed" -> "build.parseways.exceeded_max_speed"
+      statsd_entries.emplace_back(std::format("build.{}.{}", stage_name, meta[i].statsd_key + 6),
+                                  delta);
+    }
+    // update the current counters
+    // currently, all counters are incremented in a single stage
+    snapshot[i] = current;
```
Comment (nilsnolde) [inline src/mjolnir/util.cc:931]: updates the data for the caller too

File: src/mjolnir/util.cc [hunk 14]
```diff
@@ -899,5 +914,46 @@ TileManifest TileManifest::ReadFromFile(const std::string& filename) {
...
+                              config.get<uint64_t>("statsd.batch_size", 500), 0);
+  std::vector<std::string> tags;
+  auto added_tags = config.get_child_optional("statsd.tags");
+  if (added_tags) {
+    for (const auto& tag : *added_tags) {
+      tags.push_back(tag.second.data());
+    }
+  }
+  for (const auto& [key, count] : statsd_entries) {
+    client.gauge(key, count, 1.f, tags);
```
Comment (nilsnolde) [inline src/mjolnir/util.cc:950]: gauge overwrites previous metrics. grafana can alert on sudden changes.

**Comment by nilsnolde** [comment]
@kevinkreiser I changed a few things according to your comments:
- every stage emits a summary log & statsd after it's done
- removed `reset` for the singleton, was indeed unnecessary, it's only ever instantiated once per process (unless for gurka, but don't need to care about it)

and again: AI was really crap on this PR! everything done in a single context, couldn't connect the most obvious dots..

File: src/mjolnir/shortcutbuilder.cc [hunk 15]
```diff
@@ -772,16 +777,23 @@ void ShortcutBuilder::Build(const boost::property_tree::ptree& pt) {
...
+    [[maybe_unused]] uint32_t avg = sc_count ? (edge_count / sc_count) : 0;
+    LOG_INFO("Finished with " + std::to_string(sc_count) + " shortcuts superseding " +
+             std::to_string(edge_count) + " edges, average ~" + std::to_string(avg) +
              " edges per shortcut");
+    total_exceeded_max += exceeded_max;
+  }
+
+  if (total_exceeded_max > 0) {
+    LOG_WARN(std::to_string(total_exceeded_max) + " nodes exceeding max shortcut edges");
+    build_stats::get().counters[build_stats::kExceededMaxShortcutEdges] = total_exceeded_max;
```
Comment (kevinkreiser) [inline src/mjolnir/shortcutbuilder.cc]: this kind of stuff hurts a little bit, modifying internal state of a global

File: src/mjolnir/util.cc [hunk 16]
```diff
@@ -899,5 +914,44 @@ TileManifest TileManifest::ReadFromFile(const std::string& filename) {
...
+  }
+  client.flush();
```
Comment (kevinkreiser) [inline src/mjolnir/util.cc:952]: this is probably redundant right, `client` goes out of scope, i hope it flushes stats before its destructed

File: valhalla/mjolnir/util.h [hunk 17]
```diff
@@ -90,6 +92,78 @@ inline std::string to_string(BuildStage stg) {
...
+      {"build.access_tags_not_found", "edges with access tags not found"},
+      {"build.unrecognized_hov_type", "ways with unrecognized HOV type"},
+      {"build.tag_parse_error", "tag parse errors"},
+      {"build.exceeded_max_vias", "restrictions exceeding max vias"},
+      {"build.exceeded_max_shortcut_edges", "nodes exceeding max shortcut edges"},
+  };
+
+  static_assert(std::size(meta) == kCount, "build_stats::meta and counter enum are out of sync");
+
+  std::atomic<uint32_t> counters[kCount]{};
```
Comment (chrstnbwnkl) [inline valhalla/mjolnir/util.h]: I would prefer `std::array<std::atomic<uint32_t>, kCount>` over this C-style array, I had to look twice and think whether this is an atomic array or an array of atomics :smile: but maybe that's just me.

File: valhalla/mjolnir/util.h [hunk 18]
```diff
@@ -90,6 +92,78 @@ inline std::string to_string(BuildStage stg) {
...
+
+  void increment(counter c) {
+    ++counters[c];
```
Comment (chrstnbwnkl) [inline valhalla/mjolnir/util.h]: I prefer to explicitly call `std::atomic` member functions than relying on its operator overloads; also, I think in this case we don't care about a strict memory order, since it seems we will only ever look at these values from the main thread?

The ++ operator should be equal to:
```cpp 
 counters[c].fetch_add(1, std::memory_order_seq_cst);
```

But I think we can get away with: 

```cpp
 counters[c].fetch_add(1, std::memory_order_relaxed);
```
 
Probably overly pedantic, just a thought :smile:

File: valhalla/mjolnir/util.h [hunk 19]
```diff
@@ -90,6 +92,78 @@ inline std::string to_string(BuildStage stg) {
...
+      {"build.invalid_level", "ways with invalid level tags"},
+      {"build.exceeded_max_names", "edges exceeding max names"},
+      {"build.exceeded_max_shape_size", "edges exceeding max encoded shape size"},
+      {"build.exceeded_speed_limit", "edges with speed limit clamped in EdgeInfo"},
+      {"build.elevation_exceeds_diff", "edges with elevation exceeding max difference"},
+      {"build.access_tags_not_found", "edges with access tags not found"},
+      {"build.unrecognized_hov_type", "ways with unrecognized HOV type"},
+      {"build.tag_parse_error", "tag parse errors"},
+      {"build.exceeded_max_vias", "restrictions exceeding max vias"},
+      {"build.exceeded_max_shortcut_edges", "nodes exceeding max shortcut edges"},
```
Comment (chrstnbwnkl) [inline valhalla/mjolnir/util.h]: What do you think about renaming the `build` namespace to `mjolnir`? That would be in line with the other namespaces we use for tracking service stats

Comment (kevinkreiser) [reply valhalla/mjolnir/util.h]: this didnt even come to my mind but i second this!

File: src/mjolnir/shortcutbuilder.cc (hunk 15 above)
Comment (nilsnolde) [reply src/mjolnir/shortcutbuilder.cc]: urgh, I missed this one! yeah, no, we're not gonna do that:)

File: valhalla/mjolnir/util.h (hunk 17 above)
Comment (nilsnolde) [reply valhalla/mjolnir/util.h]: na I do prefer c++ abstractions too!

File: valhalla/mjolnir/util.h (hunk 18 above)
Comment (nilsnolde) [reply valhalla/mjolnir/util.h]: hm, here I find the operator overload a lot more readable tbh. IMO the benefit of `relaxed` ordering is rather negligible since we won't call this is very hot loops like edge counting or so.

File: valhalla/mjolnir/util.h (hunk 19 above)
Comment (nilsnolde) [reply valhalla/mjolnir/util.h]: yep, good point, done! removed the prefix here and add `mjolnir.` to the place where metrics are collected before sending/logging them in `mjolnir/util.cc`:

File: valhalla/mjolnir/util.h (hunk 17 above)
Comment (nilsnolde) [reply valhalla/mjolnir/util.h]: done

**Comment by nilsnolde** [comment]
I just completed a first full planet build with the extension of this PR (https://github.com/valhalla/valhalla/pull/5991). at least it proved the infrastructure is set up properly:
image

but the log stilled showed tons of warnings for:
- exceeding max density
- exceeding turn restrictions mask
- some left over speed stuff etc

there's been a few more things. I want to iterate on this a bit more before merging.

**Comment by nilsnolde** [comment]
closing in favor of https://github.com/valhalla/valhalla/pull/6004
